applications/GeoMechanicsApplication/custom_elements/transient_Pw_line_element.h

```diff
@@ -0,0 +1,419 @@
+
+        auto result = BoundedMatrix<double, TNumNodes, TNumNodes>{ZeroMatrix{TNumNodes, TNumNodes}};
+        for (unsigned int integration_point_index = 0;
+             integration_point_index < GetGeometry().IntegrationPointsNumber(GetIntegrationMethod());
+             ++integration_point_index) {
+            const auto N = Vector{row(r_N_container, integration_point_index)};
+            double BiotModulusInverse = CalculateBiotModulusInverse(rCurrentProcessInfo, integration_point_index);
+            result += -BiotModulusInverse * outer_prod(N, N) * rIntegrationCoefficients[integration_point_index];
+        }
+        return result;
+    }
+
+    array_1d<double, TNumNodes> GetNodalValuesOf(const Variable<double>& rNodalVariable) const
+    {
+        auto        result     = array_1d<double, TNumNodes>{};
+        const auto& r_geometry = GetGeometry();
+        std::transform(r_geometry.begin(), r_geometry.end(), result.begin(), [&rNodalVariable](const auto& node) {
+            return node.FastGetSolutionStepValue(rNodalVariable);
+        });
+        return result;
+    }
+
+
+    void Initialize(const ProcessInfo& rCurrentProcessInfo) override
+    {
+        const std::size_t number_integration_points = GetGeometry().IntegrationPointsNumber(GetIntegrationMethod());
+
+        if (mRetentionLawVector.size() != number_integration_points) {
+            mRetentionLawVector.resize(number_integration_points);
+        }
+        for (unsigned int i = 0; i < mRetentionLawVector.size(); ++i) {
+            mRetentionLawVector[i] = RetentionLawFactory::Clone(GetProperties());
+            mRetentionLawVector[i]->InitializeMaterial(
+                GetProperties(), GetGeometry(), row(GetGeometry().ShapeFunctionsValues(GetIntegrationMethod()), i));
+        }
+    }
+
+
+    double CalculateBiotModulusInverse(const ProcessInfo& rCurrentProcessInfo,
+                                   const unsigned int integration_point_index) const
+    {
+        const auto& r_properties = GetProperties();
+        const double biot_coefficient = r_properties[BIOT_COEFFICIENT];
+
+        double result = 0.0;
+        if (!r_properties[IGNORE_UNDRAINED]) {
+            result = (biot_coefficient - r_properties[POROSITY]) / r_properties[BULK_MODULUS_SOLID] 
+                + r_properties[POROSITY] / r_properties[BULK_MODULUS_FLUID];
+        } 
+        else {
+            result = (biot_coefficient - r_properties[POROSITY]) / r_properties[BULK_MODULUS_SOLID] 
+                + r_properties[POROSITY] / TINY;
+        }
+
+        RetentionLaw::Parameters RetentionParameters(GetProperties(), rCurrentProcessInfo);
+        double degree_of_saturation = mRetentionLawVector[integration_point_index]->CalculateSaturation(RetentionParameters);
+        double derivative_of_saturation = mRetentionLawVector[integration_point_index]->CalculateDerivativeOfSaturation(RetentionParameters);
```

Comment:
These two can be const
```suggestion
        const double degree_of_saturation = mRetentionLawVector[integration_point_index]->CalculateSaturation(RetentionParameters);
        const double derivative_of_saturation = mRetentionLawVector[integration_point_index]->CalculateDerivativeOfSaturation(RetentionParameters);
```